File: src/_score/score_module.php
<?php

namespace Materia;

class Score_Modules_LanguageWidget extends Score_Module {

	public function getLegendValue($id) {
		$legend = $this->inst->qset->data['options']['legend'];

		foreach ($legend as $item) {
			if ($id == $item['id']) return $item['name'];
		}
		return '';
	}

	public function check_answer($log)
	{
		if (isset($this->questions[$log->item_id]))
		{
			$response = json_decode($log->text);

			$answer = [];

			foreach ($this->questions[$log->item_id]->answers[0]['options']['phrase'] as $token) {
				$answer[] = [
					'value' => $token['value'],
					'legend' => $this->getLegendValue($token['legend'])
				];
			}

			$max = count($answer);

			foreach ($answer as $index => $token) {

				if (array_key_exists($index, $response)) return 0;

				if ($this->questions[$log->item_id]->options['displayPref'] == 'word') {
					if (strcmp($response[$index]->value, $token['value']) != 0 || strcmp($response[$index]->legend, $token['legend']) != 0) {
						return 0;
					}
				}
				else
				{
					if (strcmp($response[$index]->legend, $token['legend']) != 0) {
						return 0;
					}
				}
			}

			return 100;
		}
	}

	protected function details_for_question_answered($log)
	{
		$q     = $this->questions[$log->item_id];
		$score = $this->check_answer($log);

		return [
			'data' => [
				$this->get_ss_question($log, $q),
				$q->id,
				$this->get_ss_answer($log, $q),
				$this->get_ss_expected_answers($log, $q)
			],
			'data_style'    => ['question', 'question_id', 'response', 'answer'],
			'score'         => $score,
			'feedback'      => $this->get_feedback($log, $q->answers),
			'type'          => $log->type,
			'style'         => $this->get_detail_style($score),
			'tag'           => 'div',
			'symbol'        => '%',
			'graphic'       => 'score',
			'display_score' => true
		];
	}
}

File: src/_score/score_module.py
import json
from scoring.module import ScoreModule

class LanguageWidget(ScoreModule):
    def get_legend_value(self, id):
        legend = self.qset["options"]["legend"]

        for item in legend:
            if id == item["id"]:
                return item["name"]
            
        return ""
    
    def check_answer(self, log):
        q = self.get_question_by_item_id(log.item_id)
        if q:
            response = json.loads(log.text)

            answer = []

            for token in q["answers"][0]["options"]["phrase"]:
                answer.append({
                    "value": token["value"],
                    "legend": self.get_legend_value(token["legend"])
                })

            if len(answer) != len(response):
                return 0
            
            for index, token in enumerate(answer):
                if index >= len(response):
                    return 0

                if q["options"]["displayPref"] == "word":
                    if response[index]["value"] != token["value"] or response[index]["legend"] != token["legend"]:
                        return 0
                else:
                    if response[index]["legend"] != token["legend"]:
                        return 0
                    
            return 100
        
        return 0

    def details_for_question_answered(self, log):
        q = self.get_question_by_item_id(log.item_id)
        score = self.check_answer(log)

        return {
            "data": [
                self.get_ss_question(log, q),
                q["id"],
                self.get_ss_answer(log, q),
                self.get_ss_expected_answers(log, q)
            ],
            "data_style": ["question", "question_id", "response", "answer"],
            "score": score,
            "feedback": self.get_feedback(log, q["answers"]),
            "type": log.log_type,
            "style": self.get_detail_style(score),
            "tag": "div",
            "symbol": "%",
            "graphic": "score",
            "display_score": True
        }
